### PR TITLE
ADD: ConnectedComponentsLabeling Box operation

### PR DESCRIPTION
**Pull request by StRigaud**
Add the connected components labelling box to CLIc and the other operation required to make it run.

The main classes do not rely on .cl files but calls various other kernel operation:
- connected components labelling box + test
- close gaps in label map (test is missing)

To runs those two operation, the following kernels were implemented:
- sum reduction + test
- replace intensity + test
- replace intensities (test is missing)
- block enumerate (test is missing)
- set column (test is missing)
- flag existing intensities (test is missing)

This required a minor modification of some .cl filename requested [here](https://github.com/clEsperanto/clij-opencl-kernels/pull/8)

I will see to add missing tests before merge. Feedback and review are welcome. Code isn't the most clean sorry :)

Message for futur me: "wrong" data type used in kernel can lead to unstable results. Further investigation and fixing may be required.

### CODE REVIEW

**Comment by haesleinhuepf** [comment]
Hey Stephane,

wow, great work! CCL is one of the most challenging "basic" operations.  As you mentioned the missing tests: One of the tests was also missing on python side. I just added it and took the oportunity to rename the function "flag_existing_intensities" to " flag_existing_labels" because that describes better what it's doing. 
https://github.com/clEsperanto/pyclesperanto_prototype/blob/flag_existing_labels/tests/test_flag_existing_labels.py

Would you mind renaming it on CLIc side as well? 

Furthermore, if you need help with implementing tests and / or specific kernels, let me know! After implementing 400 of them, I became quite efficient. ;-)

Furthermore, I'm still searching for the student who implements those for earning money. Stay tuned!

Cheers,
Robert

**Comment by StRigaud** [comment]
No issue for the re-naming :).

However, don't If you saw yet but this PR is dependent on a .cl renaming on the CLIj side. Otherwise it will not work (hence the build fail on MacOS).
I am waiting for this and to add the missing test to merge :)

**Comment by StRigaud** [comment]
Change FlagExistingIntensitiesKernel to FlagExistingLabelsKernel.
Add the corresponding test.

The kernel is still named flag_exisinting_intensities_x.cl in clij_opencl_kernel repo. The PR and CLIc will fail until this is fix :)

**Comment by StRigaud** [comment]
also set column test is not missing anymore.

still missing:
- replace intensities (test is missing)
- block enumerate (test is missing)

**Comment by StRigaud** [comment]
All test added.

However, I have some doubt on the blocksize parameter used in BlockEnumerateKernel and in SumReductionKernel.
My blocksize value must be >= to my array size otherwise the tests fails. I wouldn't mind a double check on this.

Final check should be on the gateway method name and parameters order fitting the CLIj2 and pyclesperanto method name and parameters.

File: clic/include/CLE.h [hunk 1]
```diff
@@ -94,7 +94,13 @@ class CLE
...
     void SumXProjection(Buffer&, Buffer&);
     void SumOfAllPixels(Buffer&, Buffer&);
-    
+    void ConnectedComponentLabellingBox(Buffer&, Buffer&);
+    void ReplaceIntensity(Buffer&, Buffer&, float, float);
+    void ReplaceIntensities(Buffer&, Buffer&, Buffer&);
+    void SetColumn(Buffer&, int, float);
+    void SumReduction(Buffer&, Buffer&, int);
+    void BlockEnumerate(Buffer&, Buffer&, Buffer&, int);
+    void FlagExistingLabels(Buffer&, Buffer&);
```
Comment (haesleinhuepf) [inline clic/include/CLE.h:104]: The API looks good!

File: clic/src/cleFlagExistingLabelsKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    this->AddObject(&x, "src");
+}
+
+void FlagExistingLabelsKernel::SetOutput(Object& x)
+{
+    this->AddObject(&x, "dst");
+}
+
+void FlagExistingLabelsKernel::Execute()
+{
```
Comment (haesleinhuepf) [inline clic/src/cleFlagExistingLabelsKernel.cpp:28]: Before executing this kernel, we should `set` all pixels to 0. Background is: the kernel does not write all pixels to the vector, it only writes ones. 

See also: https://github.com/clEsperanto/pyclesperanto_prototype/blob/master/pyclesperanto_prototype/_tier2/_flag_existing_intensities.py#L14

File: clic/src/cleCloseIndexGapsInLabelMapKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    unsigned int nb_indices =  int(max_value) + 1;
+
+    unsigned int indices_dim[3] = {nb_indices, 1, 1};
+    cl_mem flaggedIndices_mem = CreateBuffer<float>(src->GetBitSize() * indices_dim[0], this->gpu);
+    Buffer* flaggedIndices  = new Buffer(flaggedIndices_mem, indices_dim, src->GetDataType());
+
+    SetKernel set(this->gpu);
+    set.SetInput(*flaggedIndices);
+    set.SetValue(0);
+    set.Execute();
```
Comment (haesleinhuepf) [inline clic/src/cleCloseIndexGapsInLabelMapKernel.cpp:66]: Setting this image to 0 should be done inside the FlagExistingLabelsKernel. We can then spare this step here (se commend further down)

Comment (haesleinhuepf) [reply clic/src/cleCloseIndexGapsInLabelMapKernel.cpp:66]: That's btw. also "wrong" in pyclesperanto. Here we see how code quality benefits from rewriting things and reviewing thm :-) I will fix this on pyclesperanto side.

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hey @StRigaud ,

first of all: Happy New Year!

> My blocksize value must be >= to my array size otherwise the tests fails.

That's suspicious. The idea of the blocksize is to be smaller than the array size. It might be my mistake again, I had this error earlier. :-(

This change:
https://github.com/clEsperanto/pyclesperanto_prototype/commit/0fcbe1ce62d769a1de2b86397b04a7967cfe8d19

Didn't make it to the clij-opencl-kernels repository:
https://github.com/clEsperanto/clij-opencl-kernels/blob/master/src/main/java/net/haesleinhuepf/clij/kernels/block_enumerate.cl#L15

I urgently need to figure out how to use git submodules and restructure how the kernels go into pyclesperanto. Stay tuned. I will do this asap.

I've left some more comments below.

Thanks for again for your great work! I'm really happy to see CLIc evolving.

File: clic/src/cleFlagExistingLabelsKernel.cpp (hunk 2 above)
Comment (StRigaud) [reply clic/src/cleFlagExistingLabelsKernel.cpp:28]: Not sure if this is relevant in CLIc. `dst` Buffer seems to be created at 0. I should verify this!

I added the `Set` to 0 anyway for security. Maybe removed if not relevant in future.

File: clic/src/cleCloseIndexGapsInLabelMapKernel.cpp (hunk 3 above)
Comment (StRigaud) [reply clic/src/cleCloseIndexGapsInLabelMapKernel.cpp:66]: `SetKernel` is removed.

**Comment by StRigaud** [comment]
Add reviews modification to the PR. No issues.

For the blocksize:
I applied the .cl correction. No visible effect.
- in SumReductionKernel -> if blocksize is smaller than array length, the test fail. sums do not correspond. 
- in BlockEnumerateKernel -> if blocksize is smaller than array length, the test fail:

blocksize = 4
source = 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0
valide  = 0, 1, 0, 2, 0, 0, 3, 4, 0, 0, 5, 0
result = 0, 1, 0, 2, 0, 0, 1, 2, 0, 0, 1, 0

Again, if blocksize is high enough, all is good

**Comment by haesleinhuepf** [comment]
Hey @StRigaud .

> * in BlockEnumerateKernel -> if blocksize is smaller than array length, the test fail:

In the python version of the block_enumerate test, the sum_reduction is used. See [here](https://github.com/clEsperanto/pyclesperanto_prototype/blob/master/tests/test_block_enumerate.py#L12) 
versus [here](https://github.com/StRigaud/CLIc_prototype/blob/master/clic/test/block_enumerate_test.cpp#L42).

I attempted to fix this and sent you a [PR](https://github.com/StRigaud/CLIc_prototype/pull/1). I can't test it (Fail to create kernel (CL_INVALID_KERNEL_NAME)) I assume this has something to do with my lack of training when it comes to git submodules...

> * in SumReductionKernel -> if blocksize is smaller than array length, the test fail. sums do not correspond.

I checked [this test](https://github.com/StRigaud/CLIc_prototype/blob/master/clic/test/sum_reduction_test.cpp#L41) and have a suspicion: The output has fixed size [1,1,1]. However, it's supposed to have size `original_width / block_size`. I just wrote [a test on python side](https://github.com/clEsperanto/pyclesperanto_prototype/blob/7fe271658a3c8c0184a54e6205f2ab807a49edeb/tests/test_sum_reduction.py#L4-L6) and [added documentation](https://github.com/clEsperanto/pyclesperanto_prototype/pull/76/files#diff-25d4c5e131617c10d4ba76e1155fbe94042a7b3ee1af5a6f85a5ac4dfacf7f36R8) to make more clear what this kernel is supposed to do.

I think these issues are related to lack of documentation caused by me. I promise to improve.

Let me know if this helps!

Cheers,
Robert

**Comment by haesleinhuepf** [comment]
Oh, and one more thing: the `SumReductionKernel` should have the name `SumReductionXKernel` as it reduces the image in X dimension. We don't have them yet, but sum_reduction_y and ..._z may make sense at some point.

**Comment by StRigaud** [comment]
I haven't seen that sum_reduction_x was used in the block_enumerate test! I need to recheck the code for both and their corresponding tests.

Will do the renaming at the same time!

**Comment by StRigaud** [comment]
Thanks for the help, all work fine.
I will try to make a cleaner test later on but all checks out!

**Review by haesleinhuepf — APPROVED** [review]

**Comment by StRigaud** [comment]
All checks out! 🥇

**Pull request merged**